files/default/slurm/pcluster_slurm_config_generator.py

```diff
@@ -77,12 +81,31 @@ def _load_cluster_config(input_file_path):
 def _get_master_config():
     return {
         "master_hostname": gethostname(),
-        "master_ip": getfqdn(),
+        "master_ip": _get_master_private_ip(),
     }
 
 
+def _get_master_private_ip():
+    """Get master private ip from EC2 metadata."""
+    try:
+        private_ip = subprocess.run(
+            "curl http://169.254.169.254/latest/meta-data/local-ipv4",
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            encoding="utf-8",
+            shell=True,
+        ).stdout
+        if private_ip:
+            return private_ip
+    except Exception as e:
+        log.error("Encountered exception when retrieving master node private IP: %s", e)
+
+    log.error("Unable to retrieve master node private IP from instance metadata, using fqdn in slurm config.")
+    return getfqdn()
```

Comment:
does it make sense to use fqdn if this fails? I'd rather fail, and eventually add a `--retry 3` flag to the curl command.